File: contest/LeetCode_Weekly_Contest_203/RabbitLee/Find_Latest_Group_of_Size_M.py
"""
back-forward: O(nlogn)
"""
class Solution:
    def findLatestStep(self, arr: List[int], m: int) -> int:
        n = len(arr)
        groups = [[1, n]]
        if m == n:
            return n
        
        for i in range(n - 1, 0, -1):
            index = self.find_group(groups, arr[i])
            group = groups[index]
            if not group[0] <= arr[i] <= group[1]:
                continue
            if arr[i] - 1 - group[0] + 1 == m or group[1] - arr[i] - 1 + 1 == m:
                return i
            new_groups = [[group[0], arr[i] - 1], [arr[i] + 1, group[1]]]
            new_groups = [group for group in new_groups if group[1] - group[0] + 1 >= m]
            groups = (groups[:index] if index > 0 else []) + new_groups + groups[index + 1:]
            if not groups:
                break
        return -1
                
    def find_group(self, groups: List[List[int]], index: int) -> int:
        left = 0
        right = len(groups) - 1
        while left < right:
            mid = (left + right) // 2
            group = groups[mid]
            if group[0] <= index <= group[1]:
                return mid
            elif index < group[0]:
                right = mid - 1
            else:
                left = mid + 1
        return left

"""
Union Find: O(n)
"""
class Solution:
    class UnionFind:
        def __init__(self, n: int):
            self.parent = list(range(n))
            self.rank = [0] * n
        def find(self, u: int):
            self.parent[u] = self.find(self.parent[u]) if self.parent[u] != u else u
            return self.parent[u]
        def union(self, u: int, v: int) -> bool:
            pu, pv = self.find(u), self.find(v)     # parent of u, v
            if pu == pv:
                return False
            if self.rank[pu] > self.rank[pv]:
                self.parent[pv] = pu
                self.rank[pu] += self.rank[pv]
            else:
                self.parent[pu] = pv
                self.rank[pv] += self.rank[pu]
            return True
    
    def findLatestStep(self, arr: List[int], m: int) -> int:
        n = len(arr)
        uf = self.UnionFind(n)
        ans = -1
        for step, position in enumerate(arr):
            position -= 1
            uf.rank[position] = 1
            for neigh in (position - 1, position + 1):
                if not 0 <= neigh < n:
                    continue
                if uf.rank[neigh] > 0:
                    if uf.rank[uf.find(neigh)] == m:
                        ans = step
                    uf.union(position, neigh)
        for position in range(n):
            if uf.rank[uf.find(position)] == m:
                ans = n
                break
        return ans

File: contest/LeetCode_Weekly_Contest_203/Find_Latest_Group_of_Size_M.java
class Solution {
    class UnionFind {
        int[] parent, size;
        public UnionFind(int n) {
            parent = new int[n];
            size = new int[n];
            for (int i = 0; i < n; i++) {
                parent[i] = i;
                size[i] = 0;
            }
        }
        int find(int u) {
            if (parent[u] == u) return u;        
            parent[u] = find(parent[u]);
            return parent[u];
        }
        boolean union(int u, int v) {
            int pu = find(u), pv = find(v);
            if (pu == pv) return false;
            if (size[pu] > size[pv]) {
                parent[pv] = pu;
                size[pu] += size[pv];
            } else {
                parent[pu] = pv;
                size[pv] += size[pu];
            }
            return true;
        }
    }
    
    public int findLatestStep(int[] arr, int m) {
        int n = arr.length;
        UnionFind uf = new UnionFind(n);
        int ans = -1;
        for (int step = 0; step < n; step++) {
            int position = arr[step] - 1;
            uf.size[position] = 1;
            int[] neighs = {position - 1, position + 1};
            for (int neigh : neighs) {
                if (neigh < 0 || neigh >= n) continue;
                if (uf.size[neigh] > 0) {
                    if (uf.size[uf.find(neigh)] == m) ans = step;
                    uf.union(position, neigh);
                }
            }
        }
        for (int i = 0; i < n; i++) {
            if (uf.size[uf.find(i)] == m) {
                ans = n;
                break;
            }
        }
        return ans;
    }
}

